src/server/visibility/read.ts

```diff
@@ -0,0 +1,125 @@
+        return {
+            type: 'SPECIAL',
+            published: visibility.published,
+            id: visibility.id,
+            regular: visibility.regularLevel.permission,
+            admin: visibility.adminLevel.permission,
+            purpose: visibility.purpose
+        }
+    }
+    return {
+        type: 'REGULAR',
+        published: visibility.published,
+        purpose: visibility.purpose,
+        id: visibility.id,
+        regular: collapseVisibilityLevel(visibility.regularLevel),
+        admin: collapseVisibilityLevel(visibility.adminLevel)
+    }
+}
+
+/**
+ * This function reads a special visibility, and creates it if it does not exist
+ * (this should not happen in production)
+ * @param specialPurpose - The purpose of the special visibility
+ * @returns - The visibility
+ */
+export async function readSpecialVisibility(specialPurpose: SpecialVisibilityPurpose): Promise<VisibilityCollapsed> {
+    const visibility = await prismaCall(() => prisma.visibility.findFirst({
+        where: { specialPurpose },
+        include: {
+            regularLevel: levelSelector,
+            adminLevel: levelSelector,
+        }
+    }))
+    if (!visibility) {
+        const created = await prismaCall(() => prisma.visibility.create({
+            data: {
+                purpose: 'SPECIAL',
+                specialPurpose,
+                published: true,
+                regularLevel: {
+                    create: {
+                        permission: SpecialVisibilityConfig[specialPurpose].regularLevel
+                    }
+                },
+                adminLevel: {
+                    create: {
+                        permission: SpecialVisibilityConfig[specialPurpose].adminLevel
+                    }
+                },
+            },
+            include: {
+                regularLevel: levelSelector,
+                adminLevel: levelSelector,
+            }
+        }))
+        return {
+            type: 'SPECIAL',
+            published: created.published,
+            id: created.id,
+            regular: created.regularLevel.permission,
+            admin: created.adminLevel.permission,
+            purpose: created.purpose
+        }
+    }
+    return {
+        type: 'SPECIAL',
+        published: visibility.published,
+        id: visibility.id,
+        regular: visibility.regularLevel.permission,
+        admin: visibility.adminLevel.permission,
+        purpose: visibility.purpose
+    }
+}
+
+/**
+ * A higher order function that includes visibility in the result of a server call
+ * @param serverCall - A function to get some data, with a relation to visibility
+ * @param getVisibilityId - A methode to retrive the visibility id from the result of the server call
+ * @returns - The data from the server call, with the visibility included
+ */
+export async function includeVisibility<T>(serverCall: () => Promise<T>, getVisibilityId: (result: T) => number) {
+    const result = await serverCall()
```

Comment:
ServerCall called without any argument?

Reply:
What's the problem? if you need an argument you just need to curry.
For example:
includeVisibility(() => readNews(id), news => news.visibilityId)

Reply:
Alternativly using bind api:
includeVisibility(readNews.bind(null, id), news => news.visibilityId)

But I tend to prefer the syntax of creating a new function inside a closure